numpy/core/src/multiarray/sequence.c

```diff
@@ -69,22 +69,67 @@ NPY_NO_EXPORT PySequenceMethods array_as_sequence = {
 static int
 array_any_nonzero(PyArrayObject *arr)
 {
-    npy_intp counter;
-    PyArrayIterObject *it;
+    PyArray_NonzeroFunc *nonzero;
+    npy_intp stride, count;
+    char *data;
     npy_bool anyTRUE = NPY_FALSE;
-
-    it = (PyArrayIterObject *)PyArray_IterNew((PyObject *)arr);
-    if (it == NULL) {
+    int needs_api = 0;
+
+    NpyIter *iter;
+    NpyIter_IterNextFunc *iternext;
+    char **dataptr;
+    npy_intp *strideptr, *innersizeptr;
+    NPY_BEGIN_THREADS_DEF;
+
+    nonzero = PyArray_DESCR(arr)->f->nonzero;
+
+    iter = NpyIter_New(arr, NPY_ITER_READONLY |
+                            NPY_ITER_EXTERNAL_LOOP |
+                            NPY_ITER_REFS_OK,
+                            NPY_KEEPORDER, NPY_NO_CASTING,
+                            NULL);
+    if (iter == NULL) {
         return anyTRUE;
     }
-    counter = it->size;
-    while (counter--) {
-        if (PyArray_DESCR(arr)->f->nonzero(it->dataptr, arr)) {
-            anyTRUE = NPY_TRUE;
-            break;
-        }
-        PyArray_ITER_NEXT(it);
+    needs_api = NpyIter_IterationNeedsAPI(iter);
+
+    /* Get the pointers for inner loop iteration */
+    iternext = NpyIter_GetIterNext(iter, NULL);
+    if (iternext == NULL) {
+        NpyIter_Deallocate(iter);
+        return -1;
     }
-    Py_DECREF(it);
-    return anyTRUE;
+
+    NPY_BEGIN_THREADS_NDITER(iter);
+
+    dataptr = NpyIter_GetDataPtrArray(iter);
+    strideptr = NpyIter_GetInnerStrideArray(iter);
+    innersizeptr = NpyIter_GetInnerLoopSizePtr(iter);
+
+    /* Iterate over all the elements to count the nonzeros */
+    do {
+        data = *dataptr;
+        stride = *strideptr;
+        count = *innersizeptr;
+
+        while (count--) {
+            if (nonzero(data, arr)) {
+                anyTRUE = NPY_TRUE;
+                break;
+            }
+            if (needs_api && PyErr_Occurred()) {
+                goto finish;
+            }
+            data += stride;
+        }
+
+    } while(iternext(iter));
+
+
+finish:
+    NPY_END_THREADS;
+
+    NpyIter_Deallocate(iter);
+
+    return PyErr_Occurred() ? -1 : anyTRUE;
```

Comment:
This still seems questionable to me, but I realize you've copied it from the old code, so perhaps it needn't be addressed here.

I'd be happier with renaming `npy_bool anyTRUE` to `int ret`, and then doing `ret = -1; goto finish;` in the error case